src/utils/XFDR.cpp

```diff
@@ -0,0 +1,1095 @@
+             current_score <= TOPPXFDR::fpnum_score_end - (TOPPXFDR::fpnum_score_step/2);
+             current_score += TOPPXFDR::fpnum_score_step)
+        {
+            UInt n_intralinks = intralinks_histogram.binValue(current_score);
+            UInt n_interlinks = interlinks_histogram.binValue(current_score);
+            UInt n_monolinks = monolinks_histogram.binValue(current_score);
+            target_counts.inc(current_score, n_intralinks + n_interlinks + n_monolinks);
+        }
+     }
+     */
+    
+
+    /** Target counting as performed by the xProphet software package
+      *
+      * @brief xprophet  method for target hits counting as implemented in xProphet
+      * @param cum_histograms Cumulative score distributions
+      */
+    void fdr_xprophet(std::map< String, Math::CumulativeHistogram<>  * > & cum_histograms,
+                      const String  & targetclass, const String & decoyclass, const String & fulldecoyclass,
+                      vector< double > & fdr, bool mono)
+    {
+      for (double current_score = this->min_score +  (TOPPXFDR::fpnum_score_step/2) ;
+           current_score <= this->max_score - (TOPPXFDR::fpnum_score_step/2);
+           current_score += TOPPXFDR::fpnum_score_step)
+      {
+        double estimated_n_decoys = cum_histograms[decoyclass]->binValue(current_score);
+        if ( ! mono)
+        {
+          estimated_n_decoys -= 2 * cum_histograms[fulldecoyclass]->binValue(current_score);
+        }
+        double n_targets = cum_histograms[targetclass]->binValue(current_score);
+        fdr.push_back(n_targets > 0 ? estimated_n_decoys / (n_targets) : 0);
+      }
+    }
+
+    /**
+     * @brief Calculates the qFDR values for the provided FDR values, assuming that the FDRs are sorted by score in the input vector
+     * @param fdr Vector with FDR values which should be used for qFDR calculation
+     * @param qfdr Result qFDR values
+     */
+    void calc_qfdr(const vector< double > & fdr, vector< double > & qfdr)
+    {
+      qfdr.resize(fdr.size());
+      for (int i = fdr.size() - 1; i > -1; --i)
+      {
+        double current_fdr = fdr[i];
+        double smallest_fdr = current_fdr;
+        for (int j = i; j > -1; j--)
+        {
+          double fdr_to_check = fdr[j];
+          if (fdr_to_check < smallest_fdr)
+          {
+            smallest_fdr = fdr_to_check;
+          }
+        }
+        qfdr[i] = smallest_fdr < current_fdr ? smallest_fdr : current_fdr;
+      }
+    }
+
+    // this function will be used to register the tool parameters
+    // it gets automatically called on tool execution
+    void registerOptionsAndFlags_()
+    {
+      // Verbose Flag
+      registerFlag_(TOPPXFDR::param_verbose, "Whether the log of information will be loud and noisy");
+
+      // File input
+      registerInputFile_(TOPPXFDR::param_in, "<file>", "", "Results in the original xquest.xml format", false);
```

Comment:
This is not a subsection. I have just defined the name of the parameter as a static member of the tool.